io-sim-classes/src/Control/Monad/Class/MonadTimer.hs

```diff
@@ -178,3 +189,37 @@ diffTimeToMicrosecondsAsInt d =
     assert (usec <= fromIntegral (maxBound :: Int)) $
     fromIntegral usec
 
+{-------------------------------------------------------------------------------
+  Timeout an action
+
+  This is based on the implementation in System.Timeout
+-------------------------------------------------------------------------------}
+
+data TimeoutException m = TimeoutException (Timeout m)
+
+instance Show (TimeoutException m) where
+  show _ = "<<timeout>>"
+
+instance Typeable m => Exception (TimeoutException m) where
+  toException   = E.asyncExceptionToException
+  fromException = E.asyncExceptionFromException
+
+-- | Generalization of 'System.Timeout.timeout' to 'MonadTimer'
+--
+-- NOTE: Like most of the timer API, this will only work with the threaded
+-- runtime. When using the non-threaded runtime, will fail with a runtime
+-- exception about a pattern match failure in "GHC.Event.Thread".
+timeoutAfter :: (MonadFork m, MonadTimer m, MonadMask m, Typeable m)
```

Comment:
Note that this function subtly differs from (without even thinking about masking):
```haskell
timeoutAfter' :: (MonadAsync m, MonadTimer m) => DiffTime -> m a -> m (Maybe a)
timeoutAfter' d act =
    either (const Nothing) Just <$> race (threadDelay d) act
```
In `timeoutAfter`, the action is executed on the same thread as the one that calls `timeoutAfter` whereas in `timeoutAfter'` the action is executed in a new background thread. In most cases, this does not matter, but `timeoutAfter` will be used in combination with the `ResourceRegistry`, which does enforce that some actions on it are called from the right thread.